File: NumberOfDiscIntersections.cpp
#include <algorithm>

struct Interval {
    long start, end;
};

bool cmp(Interval i1, Interval i2)
{
    return i1.start == i2.start ? i1.end < i2.end : i1.start < i2.start;
}

int upperBound(Interval arr[], int l, int r, long x)
    {
        int mid;

        while(r - l > 1)
        {
            mid = l + (r - l)/2;

            if(arr[mid].start > x)
                r = mid - 1;
            else
                l = mid;
        }

        return arr[r].start > x ? l : r;
    }

int solution(vector<int> &A) {
    
    int n = A.size();

    Interval arr[n];

    for(int i = 0; i < n; i++)
    {
        arr[i].start = i - (long)A[i];
        arr[i].end = i + (long)A[i];
    }

    sort(arr, arr + n, cmp);
    int count = 0;

    for(int i = 0; i < n; i++)
    {
        count += upperBound(arr, i, n-1, arr[i].end) - i;
        if(count > 10000000)
            return -1;
    }

    return count;
}

File: NumberOfDiscIntersections.java
import java.util.*;
import java.io.*;

class NumberOfDiscIntersections {

    private int upperBound(long[][] arr, int l, int r, long x)
    {
        int mid;

        while(r - l > 1)
        {
            mid = l + (r - l)/2;

            if(arr[mid][0] > x)
                r = mid - 1;
            else
                l = mid;
        }

        return arr[r][0] > x ? l : r;
    }

    public int solution(int[] A) {
        
        int n = A.length;
        long[][] arr = new long[n][2];

        for(int i = 0; i < n; i++)
        {
            arr[i][0] = i - (long)A[i];
            arr[i][1] = i + (long)A[i]; 
        }

        Arrays.sort(arr, (a, b) -> 
            a[0] == b[0] ? (a[1] > b[1] ? 1 : -1) 
            : (a[0] > b[0] ? 1 : -1)
        );

        int count = 0;

        for(int i = 0; i < n; i++)
        {
            count += upperBound(arr, i, n-1, arr[i][1]) - i;
            if(count > 10000000)
                return -1;
        }

        return count;
    }
}
